src/brain/query_clusterer.cpp

```diff
@@ -0,0 +1,122 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// query_clusterer.cpp
+//
+// Identification: src/brain/query_clusterer.cpp
+//
+// Copyright (c) 2015-2018, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "brain/query_clusterer.h"
+
+namespace peloton {
+namespace brain {
+
+void QueryClusterer::UpdateFeatures() {
+  /*
+   * Read the latest queries from server over RPC
+   * Update the feature vectors for template queries
+   * For new templates - insert into templates_ and
+   * call UpdateTemplate(fingerprint, true)
+   */
+}
+
+void QueryClusterer::UpdateTemplate(std::string fingerprint, bool is_new) {
+  auto feature = features_[fingerprint];
+  double similarity = 0.0;
+  Cluster *cluster = nullptr;
+  // TODO[Siva]: Ensure that this doesn't fail when the KDTree is empty
+  kd_tree_.GetNN(feature, cluster, similarity);
+
+  if (cluster == nullptr) {
+    cluster = new Cluster(num_features_);
+    cluster->AddTemplateAndUpdateCentroid(fingerprint, feature);
+    kd_tree_.Insert(cluster);
```

Comment:
If this is going to rebuild the tree as well, please add a TODO here. We may need to come back to optimize it. For now I think we can move on.